sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_compiler.asciidoc

```diff
@@ -461,51 +597,301 @@ is `name`.
+function `bar`.
+The host code then passes the same string (`"mykernels::bar"`) to
+`ext_oneapi_get_kernel` in order to get the `kernel` object.
+The string must have exactly the same content as the string that was used to
+construct the property, without even any whitespace differences.
+
+The application can also obtain the compiler-generated (i.e. mangled) name for
+the kernel by calling `ext_oneapi_get_raw_kernel_name` like this:
+
+```
+sycl::kernel_bundle<sycl::bundle_state::ext_oneapi_source> kb_src = /*...*/;
+
+sycl::kernel_bundle<sycl::bundle_state::executable> kb = syclex::build(kb_src,
+  syclex::properties{syclex::registered_kernel_names{"mykernels::bar"}});
+
+std::string mangled_name = kb.ext_oneapi_get_raw_kernel_name("mykernels::bar");
+```
+
+Again, the string passed to `ext_oneapi_get_raw_kernel_name` must have exactly
+the same content as the string that was used to construct the
+`registered_kernel_names` property.
+The application may also pass this compiler-generated (i.e. mangled) name to
+`ext_oneapi_get_kernel` in order to get the `kernel` object.
+
+==== Instantiating templated kernel functions
+
+The `registered_kernel_names` property can also be used to instantiate a
+kernel that is defined as a function template.
+For example, consider source code that defines a kernel function template like
+this:
+
+```
+std::string source = R"""(
+  #include <sycl/sycl.hpp>
+  namespace syclex = sycl::ext::oneapi::experimental;
+
+  template<typename T>
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY((syclex::range_kernel<1>))
+  void bartmpl(T *in, T *out) {/*...*/}
+)""";
+```
+
+The application can use the `registered_kernel_names` property to instantiate
+the template for specific template arguments.
+For example, this host code instantiates the template twice and gets a `kernel`
+object for each instantiation:
+
+```
+sycl::kernel_bundle<sycl::bundle_state::ext_oneapi_source> kb_src = /*...*/;
+
+sycl::kernel_bundle<sycl::bundle_state::executable> kb = syclex::build(kb_src,
+  syclex::properties{syclex::registered_kernel_names{{"bartmpl<float>", "bartmpl<int>"}});
+
+sycl::kernel k_float = kb.ext_oneapi_get_kernel("bartmpl<float>");
+sycl::kernel k_int = kb.ext_oneapi_get_kernel("bartmpl<int>");
+```
+
+
+== Examples
+
+=== Simple example
 
 The following example demonstrates how a SYCL application can define a kernel
 as a string and then compile and launch it.
 
 ```
 #include <sycl/sycl.hpp>
 namespace syclex = sycl::ext::oneapi::experimental;
+static constexpr size_t NUM = 1024;
 
 int main() {
   sycl::queue q;
 
-  // The source code for one or more kernels, defined in one of
-  // the supported source languages.
+  // The source code for a kernel, defined as a SYCL "free function kernel".
   std::string source = R"""(
-    /* language specific kernel source code */
+    #include <sycl/sycl.hpp>
+    namespace syclex = sycl::ext::oneapi::experimental;
+
+    extern "C"
```

Comment:
I was having trouble getting this `extern "C"` approach to work, mostly because we employ a lot of templates in the headers right now and I was trying to minimize the disruption. The demangling `get_kernel_spirv_name("iota")`  is a straightforward way of avoiding the problem.

Reply:
I don't understand your comment about `extern "C"`.  Why does it matter that we use a lot of templating in our headers?  How does this affect the user's ability to declare a function `extern "C"`?

Regarding, `get_kernel_spirv_name("iota")`, see the proposed function `ext_oneapi_get_raw_kernel_name` which provides similar functionality.